src/helpers/compositor.tsx

```diff
@@ -156,108 +158,161 @@ const ElementTree = (props: { nodeId: string }) => {
-          wrapperEl.toggleAttribute('data-drop-target-ready', false)
 
-          wrapperEl.querySelectorAll('[data-item]').forEach((x) => {
-            x.toggleAttribute('data-drag-target-active', false)
-            x.toggleAttribute('data-layout-drop-target-active', false)
-            x.toggleAttribute('data-transform-drop-target-active', false)
-            x.toggleAttribute('data-transform-drop-self-active', false)
-          })
+          const adjustPreviewPosition = () => {
+            if (!isDraggingSelf.current) return
+
+            // Runs every frame while dragging
+            dragPreviewEl.style.top = latestPointerCoords.y - 20 + 'px'
+            dragPreviewEl.style.left = latestPointerCoords.x - 20 + 'px'
+            dragPreviewEl.toggleAttribute('data-active', true)
+
+            requestAnimationFrame(() => adjustPreviewPosition())
+          }
+
+          const onGlobalPointerMove = (e: PointerEvent) => {
+            if (
+              !isDraggingSelf.current &&
+              Boolean(activePointerCoords.current)
+            ) {
+              const isDragging = Boolean(
+                Math.abs(e.clientX - activePointerCoords.current.x) >=
+                  DRAG_DISTANCE_BUFFER ||
+                  Math.abs(e.clientY - activePointerCoords.current.y) >=
+                    DRAG_DISTANCE_BUFFER,
+              )
+
+              // Runs only once when dragging starts
+              if (isDragging) {
+                setDraggingNodeId(node.id)
+                isDraggingSelf.current = isDragging
+                adjustPreviewPosition()
+              }
+            }
+            latestPointerCoords = { x: e.clientX, y: e.clientY }
+          }
+
+          onGlobalPointerMove(e)
+
+          const onGlobalPointerUp = (e: PointerEvent) => {
+            setTimeout(() => {
+              log.debug('Compositor: DragEnd', e)
+              activePointerCoords.current = undefined
+              isDraggingSelf.current = false
+              setDraggingNodeId(null)
+
+              dragPreviewEl.toggleAttribute('data-active', false)
+              wrapperEl.toggleAttribute('data-dragging', false)
+              wrapperEl.toggleAttribute('data-drop-target-ready', false)
+              wrapperEl.querySelectorAll('[data-item]').forEach((x) => {
+                x.toggleAttribute('data-drag-target-active', false)
+                x.toggleAttribute('data-layout-drop-target-active', false)
+                x.toggleAttribute('data-transform-drop-target-active', false)
+                x.toggleAttribute('data-transform-drop-self-active', false)
+              })
+            })
+
+            document.removeEventListener('pointermove', onGlobalPointerMove)
+            document.removeEventListener('pointerup', onGlobalPointerUp)
+          }
+
+          document.addEventListener('pointermove', onGlobalPointerMove)
+          document.addEventListener('pointerup', onGlobalPointerUp)
```

Comment:
This is just something to consider, but I’m concerned we might risk running into memory leaks if we don’t wrap the layout and transform drag handlers in memoization. I know it’s tricky, but it could also make it easier to track changes across render cycles. Right now, it seems harder to know what lingering references we might not be cleaning up, which could lead to a memory leak. Again, this is totally up to you; I’m not enforcing anything.

Reply:
I totally agree with your concern. truth is, there's so much inside of this file that I'd like to rework, I forced myself to keep everything as much the same as I could for this PR, since it works and is tested. 

Currently it's a pretty haphazard combination of React vs native JS, so I am particularly worried about unexpected behavior if we begin to rely on the actual React lifecycle (I've had to lift some variables out of the component scope as a workaround, so component renders are not firing reliably).

I'll keep a close eye on performance and make a ticket for optimizations if needed. Regardless, I think we need to plan for some cleanup effort the next time changes are needed in this file.